### PR TITLE
http conn manager: configure cidr_ranges for internal address

### PR DESCRIPTION
**Pull request by deveshkandpal1224**
- This PR addresses the issue highlighted here -> https://github.com/envoyproxy/envoy/issues/19844
- Adds `cidr_ranges` to http connection manager's internal address configuration API
- adds helper function to determine if cidr_ranges is empty in `cidr_ranges.cc`
- Adds unit test for same

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/filters/network/http_connection_manager/config.h [hunk 1]
```diff
@@ -99,12 +100,12 @@ class InternalAddressConfig : public Http::InternalAddressConfig {
       return unix_sockets_;
     }
 
-    // TODO(snowp): Make internal subnets configurable.
-    return Network::Utility::isInternalAddress(address);
+    return cidr_ranges_.contains(address) || Network::Utility::isInternalAddress(address);
```
Comment (ggreenway) [inline source/extensions/filters/network/http_connection_manager/config.h:103]: I think that `isInternalAddress()` should not be used if the user configured cidr ranges here. So basically if it's unconfigured, use the defaults (10.0.0.0/8 et al), and if it is configured, only use the configured cidrs.

**Review by ggreenway — CHANGES_REQUESTED** [review]
Thanks for working on this!

/wait

File: source/extensions/filters/network/http_connection_manager/config.h (hunk 1 above)
Comment (deveshkandpal1224) [reply source/extensions/filters/network/http_connection_manager/config.h]: fixed, PTAL. Thanks!

**Comment by deveshkandpal1224** [comment]
@ggreenway - thanks for reviewing this. I was trying to generate the docs as well using `./docs/build.sh` but i've been running into this issue and haven't had any luck trying to figure out how to fix it. Do you have thoughts / suggestions ? 

```

➜  envoy git:(devesh24/internal-address-enhancement) ./docs/build.sh
INFO: Build option --action_env has changed, discarding analysis cache.
ERROR: /private/var/tmp/_bazel_devesh.kandpal/eaa029ba95e9ee447c7dcec0acedcbf2/external/base_pip3_sphinx/BUILD.bazel:22:11: no such package '@base_pip3_importlib_metadata//': The repository '@base_pip3_importlib_metadata' could not be resolved: Repository '@base_pip3_importlib_metadata' is not defined and referenced by '@base_pip3_sphinx//:pkg'
ERROR: Analysis of target '//docs:html' failed; build aborted:
INFO: Elapsed time: 0.638s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (373 packages loaded, 3263 targets configured)
    currently loading: @base_pip3_uvloop// ... (8 packages)
    Fetching @local_config_xcode; Fetching the default Xcode version

```

File: test/common/http/header_map_impl_test.cc [hunk 2]
```diff
@@ -424,9 +424,8 @@ TEST_P(HeaderMapImplTest, AllInlineHeaders) {
     INLINE_REQ_RESP_STRING_HEADERS(TEST_INLINE_STRING_HEADER_FUNCS)
   }
   {
-    // No request trailer O(1) headers.
-  }
-  {
+      // No request trailer O(1) headers.
+  } {
```
Comment (zhxie) [inline test/common/http/header_map_impl_test.cc:428]: Please fix format here. Envoy provides some tool scripts and you may run `./ci/run_envoy_docker.sh './ci/do_ci.sh fix_format'` to fix it.

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 3]
```diff
@@ -202,6 +203,9 @@ message HttpConnectionManager {
 
     // Whether unix socket addresses should be considered internal.
     bool unix_sockets = 1;
+
+    // list of cidr_ranges that are treated as internal.
+    repeated config.core.v3.CidrRange cidr_ranges = 2;
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:208]: Please add a version history entry as well.

**Review by htuch — COMMENTED** [review]
/lgtm api

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto (hunk 3 above)
Comment (deveshkandpal1224) [reply api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:208]: updated `current.rst`, PTAL when you have a moment @htuch. Thanks!

**Comment by deveshkandpal1224** [comment]
@ggreenway - when you have a moment - could you please take a look at the PR ? Thanks.

File: docs/root/version_history/current.rst [hunk 4]
```diff
@@ -89,6 +89,7 @@ New Features
 * config: added new :ref:`custom config validators <config_config_validation>` to dynamically verify config updates.
 * cors: add dynamic support for headers ``access-control-allow-methods`` and ``access-control-allow-headers`` in cors.
 * http: added random_value_specifier in :ref:`weighted_clusters <envoy_v3_api_field_config.route.v3.RouteAction.weighted_clusters>` to allow random value to be specified from configuration proto.
+* http: added support for :ref:`cidr_ranges <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.InternalAddressConfig.cidr_ranges>` for configuring list of `cidr_ranges` that are considered internal.
```
Comment (htuch) [inline docs/root/version_history/current.rst:92]:
```suggestion
* http: added support for :ref:`cidr_ranges <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.InternalAddressConfig.cidr_ranges>` for configuring list of CIDR ranges that are considered internal.
```

Comment (htuch) [reply docs/root/version_history/current.rst:92]: LGTM with this change.

Comment (deveshkandpal1224) [reply docs/root/version_history/current.rst]: committed.

**Comment by deveshkandpal1224** [comment]
@ggreenway - gentle reminder, if you could take a look at this PR when you have a moment. Appreciate your time!

File: api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto [hunk 5]
```diff
@@ -202,6 +203,9 @@ message HttpConnectionManager {
 
     // Whether unix socket addresses should be considered internal.
     bool unix_sockets = 1;
+
+    // list of cidr_ranges that are treated as internal.
```
Comment (htuch) [inline api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto:207]:
```suggestion
    // List of CIDR ranges that are treated as internal.
```

**Comment by phlax** [comment]
@deveshkandpal1224 please fix DCO

**Comment by deveshkandpal1224** [comment]
closing this PR because of issue fixing DCO. Current branch has other commits from `main` which is making the commit history difficult to fix. Will create a separate PR with the changes proposed from this PR in place.

**Comment by phlax** [comment]
@deveshkandpal1224 its much better just to fix DCO on this PR and force push

the new PR has lost all the existing reviews, and makes the reviewers job harder

**Pull request closed without merging**